pytorch_lightning/plugins/training_type/ddp.py

```diff
@@ -321,43 +322,30 @@ def _register_ddp_hooks(self) -> None:
                 and isinstance(self._ddp_comm_state, post_localSGD.PostLocalSGDState)
                 and self.lightning_module.trainer.state.fn == TrainerFn.FITTING
             ):
-                self._reinit_optimizers_with_post_localSGD(self._ddp_comm_state.start_localSGD_iter)
+                self._enable_model_averaging()
 
-    def _reinit_optimizers_with_post_localSGD(self, warmup_steps: int):
-        optimizers = self.lightning_module.trainer.optimizers
+    def _enable_model_averaging(self):
         if self._model_averaging_period is None:
             raise ValueError(
                 "Post-localSGD algorithm is used, but model averaging period is not provided to DDP plugin."
             )
-        averager = averagers.PeriodicModelAverager(period=self._model_averaging_period, warmup_steps=warmup_steps)
-        for x, optimizer in enumerate(optimizers):
+        for optimizer in self.lightning_module.trainer.optimizers:
             if isinstance(optimizer, LightningOptimizer):
                 optimizer = optimizer._optimizer
 
             if (
                 isinstance(optimizer, DistributedOptimizer)
+                or isinstance(optimizer, PostLocalSGDOptimizer)
                 or isinstance(optimizer, ZeroRedundancyOptimizer)
                 or (_FAIRSCALE_AVAILABLE and isinstance(optimizer, OSS))
             ):
                 raise ValueError(
-                    f"Cannot wrap a distributed optimizer of type {optimizer.__name__} by PostLocalSGDOptimizer."
+                    f"Currently model averaging cannot work with a distributed optimizer of type {optimizer.__name__}."
                 )
 
-            if isinstance(optimizer, PostLocalSGDOptimizer):
-                continue
-
-            optim_class = type(optimizer)
-            post_localSGD_optimizer = PostLocalSGDOptimizer(
-                params=optimizer.param_groups,
-                optimizer_class=optim_class,
-                averager=averager,
-                **optimizer.defaults,
-            )
-            optimizers[x] = post_localSGD_optimizer
-            del optimizer
-        trainer = self.lightning_module.trainer
-        trainer.optimizers = optimizers
-        trainer.convert_to_lightning_optimizers()
+        self.lightning_module._model_averager = averagers.PeriodicModelAverager(
```

Comment:
Yes, this will be released as a prototype feature in PyTorch 1.10, and we only have docstring for the documentation purpose so far.

See: https://github.com/pytorch/pytorch/search?q=PeriodicModelAverager